source/common/http/http2/codec_impl.cc

```diff
@@ -348,12 +348,9 @@ MetadataDecoder& ConnectionImpl::StreamImpl::getMetadataDecoder() {
 }
 
 void ConnectionImpl::StreamImpl::onMetadataDecoded(MetadataMapPtr&& metadata_map_ptr) {
-  if (decoder_destroyed_) {
-    // nghttp2 returns error, and ConnectionManager will call resetAllStreams(). Avoid referring to
-    // decoder_.
-    return;
+  if (decoder_ != nullptr) {
```

Comment:
I guess what I really don't understand is why metadata is a special case, where after failure we'll still interact with the decoder, where we can get away with asserts on the other paths.

I'd think that either after failure, the state machine should assure that we won't keep trying to interact with the decoder, or there are no higher up guarantees and we need to do checks on all the various decoder access, or metadata really is a special case here and we should comment why :-)

Reply:
Good point!! I think metadata is not a special case. When a nghttp2 callback function fails, resetAllStream() will be called ([here](https://github.com/envoyproxy/envoy/blob/5e039a6581cf5fa1f7b70329afeb2e3d35b8189c/source/common/http/conn_manager_impl.cc#L284)). After that, ActiveStream (aka, decoder_) will be scheduled to delete. If a client keeps pushing data and before the connection is closed, code_->dispatch() will keep processing data, which may refer to decoder_.

Looks like only three callbacks can return failure: ConnectionImpl::onInvalidFrame(), ConnectionImpl::onMetadataReceived() and ConnectionImpl::onMetadataFrameComplete(). I null out decoder_ in the three functions, and check decoder_ before referring it.

Will check if response direction has the same issue.